doc/designs/event-queue.md

```diff
@@ -0,0 +1,178 @@
+Event Queue and Event Dispatching design
+========================================
+
+Background
+----------
+
+For future I/O, especially to implement QUIC, a need for a generic event
+queue and event dispatcher has been identified.
+
+General description
+-------------------
+
+The general abstract model is to view an event queue as the connecting bit
+between two parts of a pipe through which an event is passed.
+
+Events themselves are carriers of information.
+
+The event queue and event dispatching designed here is currently entirely
+internal.  Public facing functionality is expected to "translate" events
+into something that makes sense for the public functionality.
+
+A publisher / subscriber model is supported primarly.  Events are considered
+activate when added to the queue, even though dispatching them to subscribers
+may be delayed, depending on when that dispatch actually happens.
+
+For timers, it means that timer events are expected to be added to the queue
+at the time the timer is supposed to trip.  To make timer events timely,
+they should be considered very high priority.
```

Comment:
I guess I'm missing how this fits in the big picture. Do we have some documentation for that? This at least does something other than what I expected.

If I understand it probably, something external to OpenSSL should be generating events saying that a socket can be read/write or that a timer has expired. OpenSSL would then be waiting for those events and process them. The application can then wait for that event to be processed.

This seems to radically change how an application talks to OpenSSL and seems to require lots of other changes to the API that are currently unclear to me.

Reply:
This is *internal*, so events are generated internally.  It actually says so, just a few lines up:

``` text
The event queue and event dispatching designed here is currently entirely
internal.  Public facing functionality is expected to "translate" events
into something that makes sense for the public functionality.
```

(I wrote "currently" to keep it open for possibly making it into a public API some time in the future, but that's not the current intent with this design)

But, you do raise a valid point that we may not have been candid enough with the grand design plans.